### ISSUE TITLE
Use Spaceport to move cargo around fleet

### ISSUE THREAD

**Issue by MrPetovan** [endless-sky/endless-sky#10768]
### Problem Description

Endless Sky currently has per-ship cargo handling, but little controls over it with most information screens showing fleet aggregates. It is possible and painstaking to move cargo around in space by jettisoning/catching it back up, but not planet-side. In particular, when buying commodities, the default cargo hold priority is at best unclear.

## Example 1

I have a Mule (170 cargo) flagship and a Clipper (70 cargo) escort. Total capacity reported on Trading screen is the expected 240 tons. I buy 150 tons of stuff, leaving 90 tons free. In space, I destroy a ship, command my escort to pick up the floating cargo, it doesn't move as its cargo bay is apparently full.

## Example 2

I have a Mule (170 cargo) flagship and a Boxwing (80 cargo) fighter. Total cargo capacity reported on Trading screen is the expected 250 tons. I buy 150 tons of stuff, leaving 100 tons free. In space, I board a ship with my mule, now the reported free cargo capacity is 20 tons as the Boxwing's cargo bay is apparently empty.

Why is the Clipper cargo hold filled first? Why isn't the Boxwing cargo hold filled first?

### Related Issue Links

This would help address #1552

### Desired Solution

In Spaceport screen, cargo (commodities and outfits) should be able to be seen per-ship and be transferred between ships. The screen should include parked ships on the planet/station and could include the outfit storage if relevant. Maybe even a planet/station cargo storage, although this is further fetched.

I recognize the UI design aspect of this screen is tricky as it must be clear what the source cargo hold and the destination cargo hold are and it must show the available/total cargo capacity for both selected ships as well. And it needs a way to swap either target with another ship in the fleet/parked, which get progressively trickier as fleets grow in size.

### Alternative Approaches

A simpler alternative is having all player ships having access to each other cargo holds in space, maybe only within a specific distance of each other?

So harvesting and looting can continue as long as there is capacity in the fleet for it. I don't particularly enjoy this idea but it would certainly smooth out the discrepancy between the detailed cargo hold game mechanics and the general lack of visibility and control over it.

**Comment by esksp** [comment]
> In Spaceport screen, cargo (commodities and outfits) should be able to be seen per-ship and be transferred between ships.

Agree, and FYI there are two other current threads about the cargo management issue: #10599 and #10717

**Comment by MrPetovan** [comment]
Thank you for the related issue search!

Note: Tried dumping cargo above another ship of my fleet from the Player Info screen, lost a 5M outfit this way. I'm back to square one with no idea how to move cargo around my fleet. My current workaround is to maximize the cargo capacity of my flagship for a lack of better solution. 🤷‍♂️

**Comment by TomGoodIdea** [comment]
Is this a duplicate of #7501?

> Note: Tried dumping cargo above another ship of my fleet from the Player Info screen, lost a 5M outfit this way. I'm back to square one with no idea how to move cargo around my fleet. My current workaround is to maximize the cargo capacity of my flagship for a lack of better solution. 🤷‍♂️

See #9218, but I guess it won't be implemented, because plundering is already overpowered.

**Comment by esksp** [comment]
> If you got the coding skills go ahead - I know I don't. Not with this UI framework.

Sorry, I've got no coding skills _at all_ (except some vestigial Visual Basic from about 20 years ago... :-D).

**Comment by MrPetovan** [comment]
I think @esksp meant "easy and quick" for the player, not for the developer 👀

**Comment by MrPetovan** [comment]
I'm 100% with you, my own expertise is web development so I know pretty well how tricky such a screen would be but I'm not able to provide an implementation in C++ at all.

**Comment by esksp** [comment]
> I think @esksp meant "easy and quick" for the player, not for the developer

Actually I _did_ mean the developer...  After all we're talking about some array and the GUI for adding, removing and moving around pointers. Sounds pretty basic to me, but then it could just be a case of severe Dunning-Kruger... :o)

**Comment by MrPetovan** [comment]
Let's see. For such a screen, you would need at least the following controls:
- Two scrollable and selectable cargo lists, ideally presented next to each other to be able to easily compare.
- Buttons to transfer either from one cargo list to the other and vice versa, either:
    - Two buttons and a popup asking for how many to transfers like for the dump cargo feature, one for each direction.
    - Four buttons, one for 1-transfer (possibly affected by control key modifiers) and one for all-transfer, in both directions.
- Two ship lists to be able to switch the current cargo displayed on either list. The shipyard/outfitter ship list would be the obvious go-to, but it doesn't scale very well with the number of ships in a fleet. While the shipyard and the outfitter screens have infinite column height, the Spaceport screen doesn't by default. If it requires a complete separate screen like the shipyard/outfitter, then it's even more work.
- An explicit feature to handle the cargo in storage that hasn't been assigned to a ship yet. This usually happens when a ship has been sold at a shipyard but its cargo capacity was in use at the time. It shows as negative free cargo on the trading screen but would need to be specifically addressed with fine-grained cargo management.
- Checks all around to verify moved cargo doesn't exceed the target hold capacity.

It isn't trivial at all to implement, even if the data concept of transferring cargo between ships is simple to grasp. GUI is one of the hardest disciplines in software programming because of the wide problem surface. By comparison algorithmic is a walk in the park because of its limited scope and the almost absence of corner cases. GUI may not seem as complex (shallow conceptual depth), but it can quickly become very complicated because of the sheer variety of supported devices (screens of various sizes, keyboard, mouse or even touch events).

### A way out?

Maybe the cheapest way of implementing this would be to copy the shipyard/outfitter screen, but with commodities/packed outfits instead. I don't know if the underlying data bindings necessary for it exist, though. Per-ship cargo implementation exists in space, but I'm not sure it can be used as-is in such a spaceport screen. Additionally, this route would require additional graphics for tradable commodities like for minable metals.

There might also be a very good reason why cargo is abstracted when landing beyond convenience which I wouldn't be aware of.

**Comment by esksp** [comment]
> Let's see. For such a screen, you would need at least the following controls:

I would see it like that:
The **cargo management screen** is a classic 3-part panel: A bigger panel with a button bar on the bottom and a sidebar with little boxes representing your ships with just their name and "Cargo: 20/50" (for instance) on the right (or left).
If you have storage on this planet, it will appear as an additional ship (it remains also accessible in the Outfitter). Yes, since you can have hundreds of ships, this ship list needs to be scrollable, sorry.
The big panel defaults to empty when you get in (or to "all cargo"? See below).

If you want to know more details about a specific ship's cargo, you click on the relevant ship and you'll get in the big panel a list (not unlike the trading one) of all items in that ship's cargo bay, with quantity (and potentially value?). The "ships" sidebar is always there, so you can quickly look through all your ships by clicking on them in the sidebar, and having their contents appear in the big panel. Storage appears as just another ship if you have something stored (else it's invisible).

Now you want to do something about one of those cargo items? You select it, and click on one of the **"Sell"/"Store"/"Move"/"Dump"** buttons on the bottom. This pops a little choice menu up, different for each action.
Examples: for moving for instance you'll get a list, with planet storage on top, and all available ships (with their free space). Clicking on one will allow you to move the item there. Item selection included, this is 3 clicks.
BTW, why have a separate "Store" button if you can store with "Move"? Because else the _new_ player won't find it... Besides "Store" just moves the item to storage, no confirmation, so it's faster (2 clicks) for when you are certain of what you want.
For selling and dumping cargo only a confirmation is needed, that's 3 clicks.

There should be a last button I didn't mention, separate from others: A **"By Ship"/"All cargo" mode toggle**. The first mode is a view by ship, as I explained above, but in the "All cargo" mode your ships list disappears, and you get a list of _all_ cargo available on this planet, with "(cargo)" or "(storage)" after each item.
Actions work the same.
"All cargo" should be the default mode when you enter the cargo management menu. 

So, that is a quick & crude back-of-the-envelope design, there is room for improvement but I must go now (Real Life calls!)!

> GUI is one of the hardest disciplines

Agree. Though I would rephrase that as "GUI is one of the hardest disciplines _to get right_".

**Comment by xobes** [comment]
I'll volunteer to implement this if we can agree on a path forward.

**Comment by Mernom** [comment]
Here are my ideas.
1) currently, cargo priority is sorted from the top to the bottom of your active escort list, with the player ship being either first or last depending on setting.
Change it so that each escort has a number between 0 and 100 associated with it, and assign the cargo from highest number to lowest. Give ship classes sensible defaults (Heavy freighter > light freighter > utility > cruise liner > transport > superheavy > heavy war > medium war > light war > interceptor > figher > drone)

2) for the actual per ship inventory system, have it be a clone of the outfitter screen. Your variable cargo is everything not currently assigned, and you can 'install' it on ships according to their cargo capacity. You can also make 'reservations' for cargo on a ship, which will cause it to always prefer to carry that stuff instead of cargo assignment being a one time deal. Perhaps ever rebuying it like ammo currently does.

3) Perhaps also a way to mark ships as 'non cargo', which would make it so that the 'buy all' button doesn't fill them when pressed for the first time. I always sell 500-1500 freight when I take off to make room for floatsom every time, so being able to have it as a built in feature would be awesome.
